src/sets/signerSet.ts

```diff
@@ -0,0 +1,364 @@
+  removeCustody: (custodyAddress: string) => void;
+  addSigner: (signerKey: string) => void;
+  removeSigner: (signerKey: string) => void;
+};
+
+class SignerSet extends TypedEmitter<SignerSetEvents> {
+  private _custodyAdds: Map<string, CustodyAddEvent>;
+  private _custodyRemoves: Map<string, CustodyRemoveAll>;
+  private _signerAdds: Map<string, SignerAdd>;
+  private _signerRemoves: Map<string, SignerRemove>;
+
+  constructor() {
+    super();
+    this._custodyAdds = new Map();
+    this._custodyRemoves = new Map();
+    this._signerAdds = new Map();
+    this._signerRemoves = new Map();
+  }
+
+  getCustodyAddresses(): Set<string> {
+    return new Set([...this._custodyAdds.keys()]);
+  }
+
+  getDelegateSigners(): Set<string> {
+    return new Set([...this._signerAdds.keys()]);
+  }
+
+  getAllSigners(): Set<string> {
+    return new Set([...this.getCustodyAddresses(), ...this.getDelegateSigners()]);
+  }
+
+  lookup(signer: string) {
+    return this.lookupCustody(signer) || this.lookupSigner(signer);
+  }
+
+  lookupCustody(custodyAddress: string) {
+    return this._custodyAdds.get(custodyAddress);
+  }
+
+  lookupSigner(signer: string) {
+    return this._signerAdds.get(signer);
+  }
+
+  merge(message: SignerMessage | CustodyRemoveAll): Result<void, string> {
+    if (isCustodyRemoveAll(message)) {
+      return this.mergeCustodyRemoveAll(message);
+    }
+
+    if (isSignerRemove(message)) {
+      return this.mergeSignerRemove(message);
+    }
+
+    if (isSignerAdd(message)) {
+      return this.mergeSignerAdd(message);
+    }
+
+    return err('SignerSet.merge: invalid message format');
+  }
+
+  /**
+   * addCustody adds a custody address to custodyAdds set. The method follows this high-level logic:
+   * - If the new address has never been seen before, add it
+   * - If the new address is already in custodyAdds, keep the entry with a higher block number
+   * - If the new address is in custodyRemoves, find the block number of the address that removed it
+   *   and move the address to custodyAdds if the new entry has a higher block number
+   *
+   * @param event - event from Farcaster ID Registry
+   */
+  addCustody(event: CustodyAddEvent): Result<void, string> {
+    const sanitizedAddress = sanitizeSigner(event.custodyAddress);
+
+    // If custody exists in custodyAdds
+    const existingCustodyAdd = this._custodyAdds.get(sanitizedAddress);
+    if (existingCustodyAdd) {
+      // If existing block number wins (greater or equal), no-op
+      if (existingCustodyAdd.blockNumber >= event.blockNumber) return ok(undefined);
+    }
+
+    // If custody exists in custodyRemoves
+    const existingCustodyRemove = this._custodyRemoves.get(sanitizedAddress);
+    if (existingCustodyRemove) {
+      const existingCustodyAddEvent = this._custodyAdds.get(sanitizeSigner(existingCustodyRemove.signer));
+
+      if (existingCustodyAddEvent) {
+        // If existing block number wins (same or greater block), no-op
```

Comment:
more details here would be helpful to readers

```suggestion
        // If the incoming add operation adds an address that was previously removed
        // and if the address that performed the remove operation is more recent than the incoming add operation
        // do nothing
```